tools/lkl/lib/virtio.c

```diff
@@ -67,16 +68,22 @@ void virtio_req_complete(struct virtio_req *req, uint32_t len)
 	struct virtio_queue *q = req->q;
 	struct virtio_dev *dev = req->dev;
 	uint16_t idx = le16toh(q->used->idx) & (q->num - 1);
-	int send_irq = 0;
+	uint16_t new = le16toh(q->used->idx) + 1;
 
 	q->used->ring[idx].id = htole16(req->idx);
 	q->used->ring[idx].len = htole16(len);
-	if (virtio_get_used_event(q) == q->used->idx)
-		send_irq = 1;
-	q->used->idx = htole16(le16toh(q->used->idx) + 1);
-
-	if (send_irq)
+	q->used->idx = htole16(new);
+
+	/* The condition is that only trigger the irq if
+	 * used->idx >= used_event + 1 > last_used_idx_signaled
+	 * i.e., the used_event set by virtio_net driver is between used->idx
+	 * and last_used_idx_signaled.
```

Comment:
virtio_get_used_event is set by kernel driver meaning "notify me when this idx of used ring is set".
Here we use  last_used_idx_signaled to record the idx when last time irq is triggered. When used->idx >= virtio_get_used_event + 1 > last_used_idx_signaled, it means we've passed that virtio_get_used_event and it has not been notified.